File: android/app/src/main/java/de/repeatuntil/yata/domain/todolist/TodoListService.kt
package de.repeatuntil.yata.domain.todolist

import de.repeatuntil.yata.domain.common.Id
import de.repeatuntil.yata.domain.common.mapFailureToRepositoryException
import de.repeatuntil.yata.domain.todolist.TodoListServiceException.TodoListDoesNotBelongToUser
import de.repeatuntil.yata.domain.todolist.TodoListServiceException.TodoListDoesNotExistForUser
import kotlinx.coroutines.CoroutineDispatcher
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

class TodoListService(
    private val repository: TodoListRepository,
    private val ioDispatcher: CoroutineDispatcher = Dispatchers.IO
) {

    suspend fun loadTodoListForUser(userId: Id): Result<TodoList?> {
        return withContext(ioDispatcher) {
            repository.getTodoListForUser(userId)
                .mapFailureToRepositoryException()
        }
    }

    suspend fun addTodoList(todoList: TodoList): Result<Unit> {
        return withContext(ioDispatcher) {
            repository.addTodoList(todoList)
                .mapFailureToRepositoryException()
        }
    }

    suspend fun updateTodoListForUser(todoList: TodoList, userId: Id): Result<Unit> {
        return withContext(ioDispatcher) {
            return@withContext if (todoList.userId == userId) {
                val todoListExists = loadTodoListForUser(userId).getOrNull() != null
                if (todoListExists) {
                    repository.updateTodoList(todoList)
                        .mapFailureToRepositoryException()
                } else {
                    Result.failure(TodoListDoesNotExistForUser(userId))
                }
            } else {
                Result.failure(TodoListDoesNotBelongToUser(userId))
            }
        }
    }
}

sealed class TodoListServiceException(message: String) : Exception(message) {
    class TodoListDoesNotBelongToUser(userId: Id) : TodoListServiceException("TodoList does not belong to user $userId")
    class TodoListDoesNotExistForUser(userId: Id) : TodoListServiceException("TodoList does not exist for user $userId")
}

File: ios/YATA/Domain/TodoList/TodoListService.swift
//
//  TodoListService.swift
//  YATA
//
//  Created by Alpár Szotyori on 26.05.24.
//

import Foundation

class TodoListService {
    private let repository: TodoListRepository

    init(repository: TodoListRepository) {
        self.repository = repository
    }

    func addTodoList(_ todoList: TodoList) async -> Result<Void, TodoListServiceError> {
        return await repository.addTodoList(todoList)
            .mapErrorToRepositoryError()
            .wrapDomainError()
    }

    func loadTodoListForUser(_ userId: Id) async -> Result<TodoList?, TodoListServiceError> {
        return await repository.getTodoListForUser(userId)
            .mapErrorToRepositoryError()
            .wrapDomainError()
    }

    func updateTodoListForUser(_ todoList: TodoList, userId: Id) async -> Result<Void, TodoListServiceError> {
        guard todoList.userId == userId else {
            return Result.failure(.todoListDoesNotBelongToUser(userId))
        }

        let existingTodoList = try? await loadTodoListForUser(userId).get()
        if existingTodoList != nil {
            return await repository.updateTodoList(todoList)
                .mapErrorToRepositoryError()
                .wrapDomainError()
        } else {
            return Result.failure(.todoListDoesNotExistForUser(userId))
        }
    }

}

enum TodoListServiceError: Error {
    case domainError(_ error: DomainError)
    case todoListDoesNotBelongToUser(_ userId: Id)
    case todoListDoesNotExistForUser(_ userId: Id)

}

fileprivate extension Result where Failure == DomainError {

    func wrapDomainError() -> Result<Success, TodoListServiceError> {
        return mapError { error in
            .domainError(error)
        }
    }

}
